rangefeed: add data driven end to end test for rangefeeds

aliher1911 opened the pull request:
This PR adds datadriven test for rangefeeds. Test exercises rangefeed infrastructure end to end and verifies client visible behaviour.

Test env covers:
- feed creation (including multiple concurrent rangefeeds)
- data insertion
  - point keys/dels
  - range dels
  - sst imports
  - transactional commits
- checkpoint waiting
- captured data validation
- rangefeed error validation
- range splitting
- non-mvcc compliant ops (clear range)

Epic: none

Touches: #82715

Release note: None

aliher1911 committed:
rangefeed: add data driven end to end test for rangefeeds

This commit adds datadriven test for rangefeeds. Test exercises
rangefeed infrastructure end to end and verifies client visible
behaviour.
Test env covers:
- feed creation
- data inservion
- checkpoint waiting
- data validation

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111524)
<!-- Reviewable:end -->

aliher1911 committed:
rangefeed: datadriven tests add delete range

This commit adds del range capability to rangefeed datadriven tests.

Epic: none

Release note: None

aliher1911 committed:
rangefeed: add sst import support for datadriven tests

This commit adds support for add sst writes as well as setting
WithOnSSTable callback and verifying resulting sstables.

Epic: none

Release note: None

aliher1911 committed:
rangefeed: add split range support for datadriven tests

This commit adds tests that verify range split is handled by datadriven
tests.

Epic: none

Release note: None

aliher1911 committed:
rangefeed: add clear range support for datadriven tests

This commit adds clear range operation to datadriven tests together
with rangefeed error verification.

Epic: none

Release note: None

aliher1911 committed:
rangefeed: add transactional put to datadriven tests

This commit adds transactional puts to datadriven tests
to verify intent creation and resolution is handled.

Epic: none

Release note: None

erikgrinaker commented:
I was hoping we could make this a bit more automated with less boilerplate. I think it's reasonable to assume that every test step wants to wait for all pending rangefeed events and dump them. We should also support running multiple concurrent rangefeeds, to test that registration filtering logic and such works properly, as well as disconnects and reconnects.

I like the idea of assigning timestamp names instead of using explicit timestamps. But how do we represent timestamps in rangefeed output? Should we simply assign logical incremental numbers to them, in addition to marking them with the timestamp name (where one exists)?

How about something like this?

```
# Create some rangefeeds with/without initial/catchup scans (should we support initial scans here, i.e. use kvfeed?).
put k=a v=a1 ts=a1
put k=b v=b1
create-rangefeed id=A start=a end=z diff=true
create-rangefeed id=B start=a end=f startTime=a1 initial-scan=false
put k=a v=a2
---
# Wait for rangefeed checkpoints to reach latest write, buffer the events, and sort them (by ID/time?).
A: k=a v=a1 ts=1:a1 [initial]
A: k=b v=b1 ts=2 [initial]
A: k=a v=a2 ts=3 prev=a1
B: k=b v=b1 ts=2 [catchup]
B: k=a v=a2 ts=3

# The rangefeeds are still running. Split the range at f, then add some more writes.
split-range k=f
delete k=a
put k=b v=b2
put k=f v=f1
---
A: k=a v=nil ts=4 prev=a2
A: k=b v=b2 ts=5
A: k=f v=f1 ts=6
B: k=a v=nil ts=4
B: k=b v=b2 ts=5

# Stop the B rangefeed, then resume it as C, this time with diff=true.
stop-rangefeed id=B ts=stopped
put k=a v=a3
put k=f v=f2
create-rangefeed id=C start=a end=f startTime=stopped initial-scan=false
---
A: k=a v=a3 ts=6 prev=nil
A: k=f v=f2 ts=7 prev=f1
B: k=a v=a3 ts=6 prev=nil [catchup]
```

aliher1911 commented:
Making it less verbose shouldn't be hard, I like the idea. So we would always wait for the checkpoint before dumping what was received by rangefeed since last dump.

Timestamps in expectations are handled like this: `pkg/kv/kvserver/rangefeed/testdata/delete_range_single_range`
If you specify timestamp then it is checked against captured write time, otherwise it is ignored (used for sorting results though).

Initial scans are defined in client side rangefeed together with other features like frontier, sst handling etc. Since we want to support latter, I don't see why not to include scans as this is pretty trivial. And I think we should include that because we also test if various processor restarts and errors doesn't cause feeds to fail.

erikgrinaker commented:
>Making it less verbose shouldn't be hard, I like the idea. So we would always wait for the checkpoint before dumping what was received by rangefeed since last dump.

Yeah, and implicitly dump after every step. The datadriven framework doesn't support batching multiple commands in a single step though, but you can look at what the `mvcc_history_test.go` framework does to handle this. We should consider upstreaming it, since it's often useful.

>Timestamps in expectations are handled like this: pkg/kv/kvserver/rangefeed/testdata/delete_range_single_range
If you specify timestamp then it is checked against captured write time, otherwise it is ignored (used for sorting results though).

Yeah, but I think we should also dump timestamps for all keys/values in the output regardless of whether we have an explicit expectation on them, by synthesizing logical timestamps for all events. This is helpful to ensure e.g. ordering, uniform txn commit timestamps, consistent initial/catchup scan timestamps, uniform SST timestamps, consistent timestamps across rangefeeds, etc.

aliher1911 commented:
Multi-statement is a bit of a mess, but it's not too bad. Now it looks like:

```
# Test verifies that rangefeed delivers data written prior to feed creation
# and ensures that catch up timestamp is respected and treated as exclusive
# value.
# Note that dump is sorted by key and then by timestamp for readability.
# We expect data written prior to range feed start, excluding row at startTS.
put k=a v=1
put k=b v=2 ts=beforeFirst
put k=c v=3
put k=a v=4
create-feed startKey=a endKey=z startTs=beforeFirst
----
key=a, val=4, ts=ts4
key=c, val=3, ts=ts3

# Put more keys to check that
put k=a v=5
put k=b v=6
put k=c v=7
----
key=a, val=5, ts=ts5
key=b, val=6, ts=ts6
key=c, val=7, ts=ts7
```

Do you think adding timestamp to each event is worth? We have to read ts back after every write op, I don't think it changes test duration meaningfully, more like adding clutter to output.

erikgrinaker commented:
> Multi-statement is a bit of a mess

Yeah, that's why I suggested adding support for it upstream rather than hacking something together.

>Do you think adding timestamp to each event is worth?

I think so. Timestamps are important in rangefeeds, and we want to make sure they're handled correctly. Some examples:

* Does intent resolution always emit a uniform commit timestamp?
* Do initial scans use the scan timestamp or MVCC timestamp?
* How do we disambiguate events that otherwise have identical keys and values (e.g. idempotent writes)?
* Are timestamps identical across rangefeeds?
* Do SSTs use the original or ingested timestamp?
* Etc...

Sure, we can add explicit assertions for everything, but if we always emit the timestamp then it's both easier to inspect this in arbitrary tests, and we get free regression testing across the entire test suite even for stuff we don't have explicit assertions for.

aliher1911 commented on pkg/kv/kvserver/rangefeed/rangefeed_int_test.go:
Integration because it tests end to end. Test that we want to replace is called `rangefeed_external_test`.

miretskiy commented on pkg/kv/kvserver/rangefeed/rangefeed_datadriven_test.go:
nit: perhaps lets skip abbreviations?

miretskiy commented on pkg/kv/kvserver/rangefeed/rangefeed_datadriven_test.go:
```diff
@@ -0,0 +1,1020 @@
+
+var commands = map[string]command{
+	"options":     handleOptions,
+	"put":         handlePutKey,
```
nit: it would be nice to add comments on the format of these commands.
While I can certainly read the data driven files, sometimes there is a bit too much magic - e.g. put command maintains "timestamp names".  It would be nice to not have to jump through hoops to see what's what.

miretskiy commented on pkg/kv/kvserver/rangefeed/rangefeed_datadriven_test.go:
i'm a bit worried about the complexity of the rangefeed test driver code; in particular, what
happens if e.g. after sending checkpoint, rangefeed restarts (for whatever reasons -- and, btw -- it would be nice to add those restarts as a matter of test coverage) and after restart resends the same checkpoint... Then it's a race between this test completing and channel being closed twice.

I don't have a good answer here.  Perhaps taking rangefeed() method from dist_sender_rangefeed_test and foregoing the use of rangefeed library might make things simpler.  Perhaps not.  Is there anything that can be done to build up the confidence that this test isn't flaky even if some transient errors happen (and as I said above: it would be *great* to inject transient errors in the rangfeed event stream).

miretskiy reviewed (commented):
This is awesome, @aliher1911
